### PR TITLE
Fix #4732: Root the desugared call to StringContext

### LINKED ISSUE #4732: Failing to resolve to correct StringContext
```scala
object StringContext {
  s"Hello"
}
```
fails to compile with Dotty and Scalac:
```scala
-- [E050] Reference Error: tests/allan/Test.scala:49:3 -------------------------
49 |  s"Hello"
   |   ^
   |   object StringContext does not take parameters
```

This is because the code snippet desugars to
```scala
StringContext("Hello").s()
```

Is this on purpose? An easy fix would be to desugar to `scala.StringContext("Hello").s()`

### PR DESCRIPTION
**Pull request by lloydmeta**
Closes #4732

* Make desugared `s""` call `_root_.scala.StringContext` instead of just
  `StringContext`
* Adds a test

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala [hunk 1]
```diff
@@ -1109,8 +1109,13 @@ object desugar {
...
         }
-
-        Apply(Select(Apply(Ident(nme.StringContext), strs), id), elems)
+        // This is a deliberate departure from scalac, where StringContext is not rooted (See #4732)
+        val rootedStringContext = {
+          val rootPkg = Ident(nme.ROOTPKG)
+          val scalaPkg = Select(rootPkg, nme.scala_)
+          Select(scalaPkg, nme.StringContext)
+        }
+        Apply(Select(Apply(rootedStringContext, strs), id), elems)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala]: Maybe `rootedStringContext` can be replaced with `ref(defn.StringContextModule)`

Comment (allanrenucci) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala]: No, Desugar produces untyped trees

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala]: Oh, I missed that. Indore my comment

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala [hunk 2]
```diff
@@ -1109,8 +1109,13 @@ object desugar {
           case Block(Nil, expr) => expr // important for interpolated string as patterns, see i1773.scala
           case t => t
         }
-
-        Apply(Select(Apply(Ident(nme.StringContext), strs), id), elems)
+        // This is a deliberate departure from scalac, where StringContext is not rooted (See #4732)
+        val rootedStringContext = {
+          val rootPkg = Ident(nme.ROOTPKG)
+          val scalaPkg = Select(rootPkg, nme.scala_)
+          Select(scalaPkg, nme.StringContext)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala]: As a shortcut, you can use scalaDot(nme.StringContext)

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala (hunk 1 above)
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala]: As suggested by @smarter, you can rewrite as
```scala
Apply(Select(Apply(scalaDot(nme.StringContext), strs), id), elems)
```

File: compiler/src/dotty/tools/dotc/transform/localopt/StringInterpolatorOpt.scala [hunk 3]
```diff
@@ -36,7 +36,7 @@ class StringInterpolatorOpt extends MiniPhase {
     def unapply(tree: Select)(implicit ctx: Context): Boolean = {
       tree.symbol.eq(defn.StringContextModule_apply) && {
         val qualifier = tree.qualifier
-        qualifier.isInstanceOf[Ident] && qualifier.symbol.eq(defn.StringContextModule)
+        qualifier.isInstanceOf[Select] && qualifier.symbol.eq(defn.StringContextModule)
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/transform/localopt/StringInterpolatorOpt.scala]: Thinking about it, the first test might not be needed. The method could be rewritten as:
```scala
def unapply(tree: Select)(implicit ctx: Context): Boolean =
  tree.symbol.eq(defn.StringContextModule_apply) &&
  tree.qualifier.symbol.eq(defn.StringContextModule)
```

File: tests/run/rooted_stringcontext.scala [hunk 4]
```diff
@@ -0,0 +1,10 @@
+object StringContext {
+  val i = 42
+  val s = s"Answer: $i"
+}
+
+object Test {
+  def main(args: Array[String]): Unit = {
+    println(StringContext.s)
```
Comment (allanrenucci) [inline tests/run/rooted_stringcontext.scala]: I would remove the check file and simply do `assert(StringContext.s == "Answer: 42")`

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala (hunk 1 above)
Comment (lloydmeta) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala]: Good idea. Done.

File: compiler/src/dotty/tools/dotc/transform/localopt/StringInterpolatorOpt.scala (hunk 3 above)
Comment (lloydmeta) [reply compiler/src/dotty/tools/dotc/transform/localopt/StringInterpolatorOpt.scala]: Indeed, it seemed a bit fragile. Removed.

File: tests/run/rooted_stringcontext.scala (hunk 4 above)
Comment (lloydmeta) [reply tests/run/rooted_stringcontext.scala]: Done.

**Comment by lloydmeta** [comment]
@allanrenucci Just addressed all your feedback + @smarter 's suggestions; please have another look when you can 😄

**Pull request merged**